Firewall: Add STM32MP25 Resource Framework Isolation support

GseoC opened the pull request:
**This P-R adds support for the hardware firewall configuration of the STM32MP25 platform.**

The resource isolation framework (RIF) is a comprehensive set of hardware blocks
designed to enforce and manage the isolation of STM32MP25 hardware resources, like
memories and peripherals.

Application software can use RIF to assign hardware resources to dedicated processing
environments that are isolated from each other using hardware isolation boundaries. Within
these environments masters are assigned to a security domain and can be secure,
non-secure, running privileged or unprivileged code, executing from a CPU, an MCU, or
a complex hardware subsystem like a multimedia engine.

Several hardware firewalls are part of the RIF. The one introduced in this P-R is the RIFSC.
The RIFSC (RIF security controller) is composed of three sets of registers, each managing a
specific set of hardware resources:

- RISC registers associated with RISUP logic (resource isolation slave unit for
peripherals), assign all non RIF-aware peripherals to zero, one or any security domains
(secure, privilege, compartment).
- RIMC registers: associated with RIMU logic (resource isolation master unit for address
space protection), assign all non RIF-aware bus master to one security domain by
setting secure, privileged, and compartment information on the system bus.
Alternatively, the RISUP logic controlling the slave port access to a peripheral can
assign target bus attributes to this peripheral master port (supported attribute: CID).
- RISC registers associated with RISAL logic (resource isolation slave unit for address
space - Lite version), assign address space subregions to one security domain (secure,
privilege, compartment).

RISALs are not supported yet in this P-R.

I've created firewall folders to store all firewall-related content.

There are some checkpatch warnings for macros and traces that I cannot handle.

etienne-lms commented on core/drivers/firewall/stm32_rifsc.c:
```diff
@@ -0,0 +1,563 @@
+		DMSG("No global lock on RIF configuration");
+		return TEE_SUCCESS;
+	}
+
```
You could remove this empty line as this assert is straight related to the instruction right above.

etienne-lms commented on core/drivers/firewall/stm32_rif.c:
```diff
@@ -0,0 +1,141 @@
+	/*
+	 * Make sure we set the bits in the right sec and priv conf register.
+	 * This is done to support IPS having more than 32 channels.
+	 */
```
I don't think the 1st sentence is really useful. I would replace this with:
`/* Some peripherals have more than 32 RIF channels */`

etienne-lms commented on core/drivers/firewall/stm32_rif.c:
```diff
@@ -0,0 +1,141 @@
+	/* CID configuration */
+	conf_data->cid_confs[channel_id] = rif_conf & cidcfdg_conf_mask;
+
+	/* This resource will be configured */
```
`/* Store that this RIF resource is to be configured */`

etienne-lms reviewed (commented):
This is my first rounds of review for these changes. I'll will likely come back with other comments.

Note for commit `drivers: firewall: add stm32_rif driver for common RIF features`: You can remove the reference to the checkpatch report from the commit log.

GseoC commented on core/drivers/firewall/stm32_rif.c:
```diff
@@ -0,0 +1,141 @@
+	io_setbits32(addr, _SEMCR_MUTEX);
+
+	/* Check that cortex A has the semaphore */
+	if (stm32_rif_is_semaphore_available(addr) ||
```
This functions returns TEE_ERROR_ACCESS_DENIED when the semaphore is free to take after trying to take it OR the semaphore has been taken by someone else. I think it's fine and the comment above is correct: We check that the cortex-A has indeed taken the semaphore

etienne-lms replied:
When a function label is of the form `xxx_is_yyy()` I would expect it implicitly means it returns true when _xxx is yyy_  and false when _xxx is not yyy_. It's handy when used with `if (xxx_is_yyy()) { ... }`. I still suggest to rename the function or make it return a boolean value.

GseoC replied:
I can rename the function to `stm32_rif_semaphore_is_available()` but I insist that the function returns true when the semaphore is available :)

etienne-lms replied:
This make this implementation to sound strange:
```c
if (semaphore-is-available || something-else)
    return ERROR-ACCESS-DENIED
```
I would rather expect:
```c
if (!semaphore-is-available || something-else)
    return ERROR-ACCESS-DENIED
```
Maybe I am missing something.

GseoC commented on core/include/dt-bindings/firewall/stm32mp25-rif.h:
```diff
@@ -0,0 +1,91 @@
+
+/* RIF semaphore list */
+#define EMPTY_SEMWL		0x0
+#ifdef __ASSEMBLER__
```
BIT is available and is fine IMO

GseoC commented on core/include/dt-bindings/firewall/stm32mp25-rif.h:
```diff
@@ -0,0 +1,91 @@
+
+/* Used when a field in a macro has no impact */
+#define RIF_UNUSED		0x0
+
```
There are worth being defined indeed. When we'll have a complete RIF configuration device tree, working without explicit macros makes device tree reading impossible. RIF configuration will be heavy and needs to be somewhat "user friendly"

etienne-lms commented on core/drivers/firewall/stm32_rifsc.c:
```diff
@@ -0,0 +1,563 @@
+	int i = 0;
+	int len = 0;
+
+	conf_list = (const struct dt_id_attr *)fdt_getprop(fdt, node,
```
Sorry I meant cast not needed.

GseoC replied:
Reworked

GseoC commented on core/drivers/firewall/stm32_rifsc.c:
```diff
@@ -0,0 +1,563 @@
+					 (field ## _SHIFT))
+
+struct rifsc_driver_data {
+	uint32_t version;
```
Ok, changed the trace at probe as well

GseoC commented on core/drivers/firewall/stm32_rifsc.c:
```diff
@@ -0,0 +1,563 @@
+	bool priv_en;
+	uint8_t nb_rimu;
+	uint8_t nb_risup;
+	uint8_t nb_risal;
```
I'd rather keep these fields linked to the register fields if that's ok

GseoC commented on core/drivers/firewall/stm32_rifsc.c:
```diff
@@ -0,0 +1,563 @@
+};
+
+struct dt_id_attr {
+	/* The effective size of the array is meaningless here */
```
Done

GseoC commented:
Hi @jforissier , is the build issue the same as #6669 ? I can't find the explanation

GseoC commented:
The checkpatch error should disappear when squashing.

The warnings/check are concerning:
- Traces (that should be OK AFAICT)

- CHECK: Prefer using the BIT macro
#1013: FILE: core/include/dt-bindings/firewall/stm32mp25-rif.h:22:
+#define RIF_CID0_BF	 
=> It is a binding file that uses `#ifdef __ASSEMBLER__` to discriminate code and device tree use

- CHECK: Macro argument reuse 'cidcfgr' - possible side-effects?
#259: FILE: core/include/drivers/stm32_rif.h:53:
+#define SEM_MODE_INCORRECT(cidcfgr)	(!((cidcfgr) & _CIDCFGR_CFEN) ||  \
+					 !((cidcfgr) & _CIDCFGR_SEMEN) || \
+					 (!((cidcfgr) &			  \
+					    _CIDCFGR_SEMWL(RIF_CID1)) &&  \
+					  ((cidcfgr) & _CIDCFGR_SEMEN)))
=> I don't find the issue. There isn't IMO but I may be mistaken

jforissier commented:
> Hi @jforissier , is the build issue the same as #6669 ? I can't find the explanation

Yes:
```
2024-02-05T12:30:39.7880398Z rsync: [receiver] write failed on "/__w/optee_os/optee_repo_qemu_v8/out-br/build/buildroot-fs/tar/target/usr/share/qemu/edk2-arm-code.fd": No space left on device (28)
```

jforissier commented:
- The long trace is OK.
- The flexible size advice may be a good one, have you considered it?
- The BIT() macro should probably be compatible with DT. Isn't it already? Would it be difficult to make it so?
- `SEM_MODE_INCORRECT()` should probably be a static inline function (`sem_mode_incorrect()`). They are almost always better IMO (except for trivial definitions, and of course when token name manipulations are required).

GseoC commented:
> * The long trace is OK.
> * The flexible size advice may be a good one, have you considered it?
I've completely removed it in a patch tagged as [Review], it should be fine after a squash
> * The BIT() macro should probably be compatible with DT. Isn't it already? Would it be difficult to make it so?
Unfortunately I'm not quite sure now on why the BIT macro isn't replaces by it's value in the device tree. Extract from the pre-dts file:
```
  (((89) << 24) | ((0x0) << 16) | ((0x0) << 10) | ((0x0) << 9) | ((0x0) << 8) | ((0x0) << 4) | ((0x0) << 1) | ((0x0) << 0))
  (((90) << 24) | ((0x0) << 16) | ((0x0) << 10) | ((0x0) << 9) | ((0x0) << 8) | ((0x0) << 4) | ((0x0) << 1) | ((0x0) << 0))
  (((92) << 24) | ((BIT(0x1)|BIT(0x2)) << 16) | ((0x1) << 10) | ((0x1) << 9) | ((0x1) << 8) | ((0x0) << 4) | ((0x1) << 1) | ((0x1) << 0))
  (((93) << 24) | ((BIT(0x1)|BIT(0x2)) << 16) | ((0x1) << 10) | ((0x1) << 9) | ((0x1) << 8) | ((0x0) << 4) | ((0x1) << 1) | ((0x1) << 0))
  (((94) << 24) | ((BIT(0x1)|BIT(0x2)) << 16) | ((0x1) << 10) | ((0x1) << 9) | ((0x1) << 8) | ((0x0) << 4) | ((0x1) << 1) | ((0x1) << 0))
```
that leads to 
```
FATAL ERROR: Unable to parse input tree
make: *** [core/sub.mk:64: out/arm-plat-stm32mp2/core/arch/arm/dts/stm32mp257f-ev1.dtb] Error 1
make: *** Waiting for unfinished jobs....
```

I don't know why the pre-processing is not done here

> * `SEM_MODE_INCORRECT()` should probably be a static inline function (`sem_mode_incorrect()`). They are almost always better IMO (except for trivial definitions, and of course when token name manipulations are required).
Sure, already done in a review patch (Forgot I did it sorry). So it should be fine with a squash as well

GseoC commented:
> > * The long trace is OK.
> > * The flexible size advice may be a good one, have you considered it?
> 
> I've completely removed it in a patch tagged as [Review], it should be fine after a squash
> 
> > * The BIT() macro should probably be compatible with DT. Isn't it already? Would it be difficult to make it so?
> 
> Unfortunately I'm not quite sure now on why the BIT macro isn't replaced by it's value in the device tree. Extract from the pre-dts file using the BIT macro:
> 
> ```
>   (((89) << 24) | ((0x0) << 16) | ((0x0) << 10) | ((0x0) << 9) | ((0x0) << 8) | ((0x0) << 4) | ((0x0) << 1) | ((0x0) << 0))
>   (((90) << 24) | ((0x0) << 16) | ((0x0) << 10) | ((0x0) << 9) | ((0x0) << 8) | ((0x0) << 4) | ((0x0) << 1) | ((0x0) << 0))
>   (((92) << 24) | ((BIT(0x1)|BIT(0x2)) << 16) | ((0x1) << 10) | ((0x1) << 9) | ((0x1) << 8) | ((0x0) << 4) | ((0x1) << 1) | ((0x1) << 0))
>   (((93) << 24) | ((BIT(0x1)|BIT(0x2)) << 16) | ((0x1) << 10) | ((0x1) << 9) | ((0x1) << 8) | ((0x0) << 4) | ((0x1) << 1) | ((0x1) << 0))
>   (((94) << 24) | ((BIT(0x1)|BIT(0x2)) << 16) | ((0x1) << 10) | ((0x1) << 9) | ((0x1) << 8) | ((0x0) << 4) | ((0x1) << 1) | ((0x1) << 0))
> ```
> 
> that leads to
> 
> ```
> FATAL ERROR: Unable to parse input tree
> make: *** [core/sub.mk:64: out/arm-plat-stm32mp2/core/arch/arm/dts/stm32mp257f-ev1.dtb] Error 1
> make: *** Waiting for unfinished jobs....
> ```
> 
> I don't know why the pre-processing is not done here
> 
> > * `SEM_MODE_INCORRECT()` should probably be a static inline function (`sem_mode_incorrect()`). They are almost always better IMO (except for trivial definitions, and of course when token name manipulations are required).
> 
> Sure, already done in a review patch (Forgot I did it sorry). So it should be fine with a squash as well

Link to similar debate:
#6547

jforissier commented:
> > > * The BIT() macro should probably be compatible with DT. Isn't it already? Would it be difficult to make it so?
> > 
> > 
> > Unfortunately I'm not quite sure now on why the BIT macro isn't replaced by it's value in the device tree. Extract from the pre-dts file using the BIT macro:
> > ```
> >   (((89) << 24) | ((0x0) << 16) | ((0x0) << 10) | ((0x0) << 9) | ((0x0) << 8) | ((0x0) << 4) | ((0x0) << 1) | ((0x0) << 0))
> >   (((90) << 24) | ((0x0) << 16) | ((0x0) << 10) | ((0x0) << 9) | ((0x0) << 8) | ((0x0) << 4) | ((0x0) << 1) | ((0x0) << 0))
> >   (((92) << 24) | ((BIT(0x1)|BIT(0x2)) << 16) | ((0x1) << 10) | ((0x1) << 9) | ((0x1) << 8) | ((0x0) << 4) | ((0x1) << 1) | ((0x1) << 0))
> >   (((93) << 24) | ((BIT(0x1)|BIT(0x2)) << 16) | ((0x1) << 10) | ((0x1) << 9) | ((0x1) << 8) | ((0x0) << 4) | ((0x1) << 1) | ((0x1) << 0))
> >   (((94) << 24) | ((BIT(0x1)|BIT(0x2)) << 16) | ((0x1) << 10) | ((0x1) << 9) | ((0x1) << 8) | ((0x0) << 4) | ((0x1) << 1) | ((0x1) << 0))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > that leads to
> > ```
> > FATAL ERROR: Unable to parse input tree
> > make: *** [core/sub.mk:64: out/arm-plat-stm32mp2/core/arch/arm/dts/stm32mp257f-ev1.dtb] Error 1
> > make: *** Waiting for unfinished jobs....
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I don't know why the pre-processing is not done here
> 
> Link to similar debate: #6547

Yes and in fact my comment was not properly addressed (https://github.com/OP-TEE/optee_os/pull/6547#pullrequestreview-1784353942).

Building with `V=1` should give you more information on which commands are run and therefore why the preprocessing is not done.

GseoC commented:
> > > > * The BIT() macro should probably be compatible with DT. Isn't it already? Would it be difficult to make it so?
> > > 
> > > 
> > > Unfortunately I'm not quite sure now on why the BIT macro isn't replaced by it's value in the device tree. Extract from the pre-dts file using the BIT macro:
> > > ```
> > >   (((89) << 24) | ((0x0) << 16) | ((0x0) << 10) | ((0x0) << 9) | ((0x0) << 8) | ((0x0) << 4) | ((0x0) << 1) | ((0x0) << 0))
> > >   (((90) << 24) | ((0x0) << 16) | ((0x0) << 10) | ((0x0) << 9) | ((0x0) << 8) | ((0x0) << 4) | ((0x0) << 1) | ((0x0) << 0))
> > >   (((92) << 24) | ((BIT(0x1)|BIT(0x2)) << 16) | ((0x1) << 10) | ((0x1) << 9) | ((0x1) << 8) | ((0x0) << 4) | ((0x1) << 1) | ((0x1) << 0))
> > >   (((93) << 24) | ((BIT(0x1)|BIT(0x2)) << 16) | ((0x1) << 10) | ((0x1) << 9) | ((0x1) << 8) | ((0x0) << 4) | ((0x1) << 1) | ((0x1) << 0))
> > >   (((94) << 24) | ((BIT(0x1)|BIT(0x2)) << 16) | ((0x1) << 10) | ((0x1) << 9) | ((0x1) << 8) | ((0x0) << 4) | ((0x1) << 1) | ((0x1) << 0))
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > that leads to
> > > ```
> > > FATAL ERROR: Unable to parse input tree
> > > make: *** [core/sub.mk:64: out/arm-plat-stm32mp2/core/arch/arm/dts/stm32mp257f-ev1.dtb] Error 1
> > > make: *** Waiting for unfinished jobs....
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I don't know why the pre-processing is not done here
> > 
> > 
> > Link to similar debate: #6547
> 
> Yes and in fact my comment was not properly addressed ([#6547 (review)](https://github.com/OP-TEE/optee_os/pull/6547#pullrequestreview-1784353942)).
> 
> Building with `V=1` should give you more information on which commands are run and therefore why the preprocessing is not done.

Actually I think we won't be able to use the BIT macro.
BIT() is defined in an OP-TEE lib that can't be used to compile device trees. I don't think we want to define the BIT macro with a define in a DTS file.
Shall we drop this?

GseoC commented:
FYI, the only hit in the kernel I found is: `arch/arm/boot/dts/ti/omap/omap3-gta04.dtsi` that defines the macro right above the node

jforissier commented:
> BIT() is defined in an OP-TEE lib that can't be used to compile device trees.
[...]
> FYI, the only hit in the kernel I found is: `arch/arm/boot/dts/ti/omap/omap3-gta04.dtsi` that defines the macro right above the node

Hmmm OK. Then I believe you could do the same. And in a later patch we might consider moving `BIT()` out of `libutils/ext/include/util.h` into its own header file that the DT pre-processing command would include with `-include`. But it's probably not worth it until we have several users.

etienne-lms commented on core/include/drivers/stm32_rif.h:
```diff
@@ -79,26 +62,49 @@ struct rif_conf_data {
+}
+
+/* stm32_rifsc_check_tdcid - This function checks if the execution context is
+ *			     TDCID or not.
```
```c
/*
 * stm32_rifsc_check_tdcid() - Check if the execution context is TDCID or not
 * ...
```

etienne-lms commented on core/include/drivers/stm32_rif.h:
```diff
@@ -79,26 +62,49 @@ struct rif_conf_data {
 /*
- * Check every possible configuration where accessing RIF
- * is possible :
+ * stm32_rif_check_access - Check every possible configuration where accessing
```
Append parentheses to function label
 s/stm32_rif_check_access/stm32_rif_check_access()/

Ditto at line 129.

etienne-lms commented on core/include/drivers/stm32_rif.h:
```diff
@@ -108,57 +114,83 @@ static inline TEE_Result stm32_rifsc_check_tdcid(bool *tdcid_state)
+/*
+ * stm32_rif_semaphore_is_available() - Checks if the _SEMCR_MUTEX bit is set
+ *
+ * @addr:	Address of the register to read from
```
Prefer use a single space char instead of the inner inner tab.
Ditto below

(but that's not a strong opinion)

etienne-lms commented on core/drivers/firewall/stm32_rif.c:
```diff
@@ -0,0 +1,126 @@
+
+	io_clrbits32(addr, _SEMCR_MUTEX);
+
+	/* Ok if another compartment takes the semaphore before the check */
```
I think this inline should rahter say `/* Check that current context no more owns the semaphore */`

GseoC replied:
s/context/compartment but yes

etienne-lms commented on core/drivers/firewall/stm32_rifsc.c:
```diff
@@ -0,0 +1,554 @@
+	if (res)
+		return res;
+
+	res = stm32_rifsc_dt_conf_rimu(fdt, node, pdata);
```
nitpicking: `return stm32_rifsc_dt_conf_rimu(fdt, node, pdata);`

etienne-lms commented on core/include/drivers/stm32_rif.h:
```diff
@@ -0,0 +1,212 @@
+}
+
+static inline TEE_Result
+		stm32_rif_acquire_semaphore(vaddr_t addr __unused,
```
Remove 2 tabs. Here and at lines below.

etienne-lms commented on core/include/drivers/stm32_rifsc.h:
```diff
@@ -0,0 +1,25 @@
+#include <types_ext.h>
+#include <util.h>
+
+struct risup_cfg {
```
Are these 2 structures to be used outside _stm32_rifsc.c_?
If not, I would suggest to defined fine only in the scope of this source file.

GseoC replied:
Right, I'll remove the header file

GseoC commented on core/drivers/firewall/stm32_rif.c:
```diff
@@ -118,22 +110,15 @@ TEE_Result stm32_rif_acquire_semaphore(vaddr_t addr,
-	assert(msb_nb_cid_supp <= MAX_CID_BITFIELD);
-
-	if (stm32_rif_is_semaphore_available(addr))
+	if (stm32_rif_semaphore_is_available(addr))
```
A CPU might not be alone in a compartment.
While I agree stm32_rif_release_semaphore() should be released after stm32_rif_acquire_semaphore() and no other software component in the same CID should access the same resource, it can happen.
If you agree, I'd prefer to keep it this way

etienne-lms commented on core/include/drivers/stm32_rif.h:
```diff
@@ -0,0 +1,212 @@
+ *  is disabled, and static CID corresponds to Cortex A35
+ * -When CID filtering is enabled on a controller, semaphore
+ *  is enabled and is not taken, and Cortex A35 CID is
+ *  white-listed
```
s/CID is white-listed/CID is authorized/g

GseoC commented on core/include/drivers/stm32_rif.h:
```diff
@@ -79,26 +62,49 @@ struct rif_conf_data {
+	    (cidcfgr & _CIDCFGR_SEMWL(wl)))
+		return true;
+
+	return false;
```
Yep, better

etienne-lms commented on core/include/drivers/stm32_rif.h:
```diff
@@ -0,0 +1,212 @@
+ *			    RIF is possible:
+ * -When no CID filtering is enabled on a controller
+ * -When CID filtering is enabled on a controller, semaphore
+ *  is disabled, and static CID corresponds to Cortex A35
```
Suggestions:
s/Cortex A35/Cortex-A/g
s/-When/- When/g (add a space char)

etienne-lms commented on core/include/drivers/stm32_rif.h:
```diff
@@ -0,0 +1,212 @@
+ * stm32_rif_parse_cfg - Parse RIF config from Device Tree extracted information
+ *
+ * @rif_conf: Configuration read in the device tree
+ * @conf_data: Buffer containing the RIF configuration to apply for an IP
```
s/IP/peripheral/g

etienne-lms commented on core/include/drivers/stm32_rif.h:
```diff
@@ -0,0 +1,212 @@
+/* SPDX-License-Identifier: (GPL-2.0 OR BSD-3-Clause) */
```
I don't think these header files require dual licensing, it is specific to OP-TEE OS.
(contrary to DT bindings header files)

GseoC replied:
Right, I'll rework these and remove the useless header file

GseoC commented on core/drivers/firewall/stm32_rifsc.c:
```diff
@@ -0,0 +1,554 @@
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	for (i = 0; i < pdata->nrisup; i++) {
+		uint32_t value = fdt32_to_cpu(*conf_list);
```
I broke this, I'll change to fdt32_to_cpu(conf_list[i])

GseoC commented:
Updated with comments addressed

etienne-lms commented on core/drivers/firewall/stm32_rifsc.c:
```diff
@@ -0,0 +1,563 @@
+		risup->id = _RIF_FLD_GET(RIFSC_RISC_PER_ID, value);
+		risup->sec = (bool)_RIF_FLD_GET(RIFSC_RISC_SEC, value);
+		risup->priv = (bool)_RIF_FLD_GET(RIFSC_RISC_PRIV, value);
+		risup->lock = (bool)_RIF_FLD_GET(RIFSC_RISC_LOCK, value);
```
nitpicking: for consistency with lines 156/158, use the same pattern at those places.

etienne-lms reviewed (commented):
I'm fin you squash the fixup commits.

etienne-lms commented:
> @etienne-lms, are you ok with me squashing the review patch so I can launch a new CI round?

Ok, I'm fine you squash all the fixup commits, including any fixes for the latest review comments I posted.
I'll go for another review round afterward.

GseoC commented:
Updated with comments addressed and review patches squashed.

GseoC commented:
I guess there is some CI issues ATM so I'm not too concerned for now with the build errors.

The checkpatch warning is due do a long trace

jforissier commented:
> I guess there is some CI issues ATM so I'm not too concerned for now with the build errors.

The Xen CI should be fixed by https://github.com/OP-TEE/build/pull/734. I am restarting the build.
Let's ignore IBART which probably needs some local action on the test host.

> The checkpatch warning is due to a long trace

That's OK

etienne-lms commented on core/include/drivers/stm32_rif.h:
```diff
@@ -0,0 +1,188 @@
+ *
+ * @cidcfgr: CIDCFGR configuration register value
+ * @semcr: SEMCR register value
+ * @nb_cid_supp: Number of CID supported
```
for consistency with description of the other functions:
`@nb_cid_supp: Number of supported CID for the peripheral`

etienne-lms commented on core/include/dt-bindings/firewall/stm32mp25-rif.h:
```diff
@@ -0,0 +1,92 @@
+#define RIF_UNUSED		0x0
+
+/* Most below macros aim to ease DTS files readability */
+#define RIF_EXTI1_RESOURCE(x)   (x)
```
replace all space char indentation with tabulation.

etienne-lms commented on core/drivers/firewall/stm32_rifsc.c:
```diff
@@ -0,0 +1,563 @@
+		risup->pm_sem = semcfgr & _SEMCR_MUTEX &&
+				((semcfgr & _SEMCR_SEMCID_MASK) >>
+				 _SEMCR_SEMCID_SHIFT == RIF_CID1) ?
+				true : false;
```
can remove `? true : false`, implicit.

GseoC replied:
Parentheses were wrong as well!

etienne-lms commented on core/drivers/firewall/stm32_rifsc.c:
```diff
@@ -0,0 +1,563 @@
+		register_pm_core_service_cb(stm32_rifsc_sem_pm, NULL,
+					    "stm32-rifsc-semaphores");
+
+	return res;
```
`return TEE_SUCCESS`

etienne-lms reviewed (commented):
Comments for commit "drivers: firewall: add stm32_rif driver for common RIF features".

Minor comments for commit "drivers: add stm32 RIFSC support".

`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits
"dt-bindings: firewall: add RIF bindings" (with comment on s/space char/tab/g addressed),
"dt-bindings: add RIF to default bindings config for stm32mp25",
"plat-stm32mp2: conf: support RIF driver",
"dt-bindings: add RIFSC bindings",
"dt-bindings: add RIFSC to default bindings config for STM32MP25",
"plat-stm32mp2: conf: enable RIFSC driver" and
"dts: stm32: Add RIFSC configuration support for stm32mp257f-ev1" and 
"dts: stm32: add RIFSC compatible to RIFSC node in stm32mp251.dtsi".

GseoC commented on core/include/drivers/stm32_rif.h:
```diff
@@ -0,0 +1,188 @@
+#define __DRIVERS_STM32_RIF_H
+
+#include <dt-bindings/firewall/stm32mp25-rif.h>
+#include <io.h>
```
dt-bindings/firewall/stm32mp25-rif.h is needed :) Code below use macros such as RIF_CID1

I will switch `#include <io.h>` for` #include <types_ext.h>` as we use vaddr_t type below

GseoC commented:
Updated with comments addressed and tags applied. Please notice I've added extra changes on 
`[Review] drivers: firewall: add stm32_rif driver for common RIF features`

etienne-lms commented on core/drivers/firewall/stm32_rif.c:
```diff
@@ -0,0 +1,124 @@
+				   __builtin_clz(nb_cid_supp | 1) - 1;
+
+	/* SCID bitfield highend can't be > SCID_SHIFT + MAX_CID_BITFIELD */
+	assert(msb_nb_cid_supp < MAX_CID_BITFIELD);
```
Could you check this max max. It seems the max number of CIDs is 8 (ID 0 to 7), hence 3 bits are needed. Maybe replace `<` with `<=`.

etienne-lms commented on core/drivers/firewall/stm32_rif.c:
```diff
@@ -0,0 +1,124 @@
+#include <kernel/panic.h>
+#include <mm/core_memprot.h>
+
+static uint32_t get_scid_mask(unsigned int nb_cid_supp)
```
I got a bit confused with this. scid refers to semaphore CID or static CID.
I think a brief inline description of this local function is welcome.

GseoC commented on core/drivers/firewall/stm32_rif.c:
```diff
@@ -0,0 +1,126 @@
+	uint32_t semwl_mask = 0;
+	unsigned int conf_index = 0;
+
+	semwl_mask = GENMASK_32(SEMWL_SHIFT + nb_cid_supp, SEMWL_SHIFT);
```
Indeed, thanks

GseoC commented on core/drivers/firewall/stm32_rifsc.c:
```diff
@@ -0,0 +1,563 @@
+#define MAX_RISUP			U(128)
+
+#define _RIF_FLD_PREP(field, value) \
+	(((uint32_t)(value) << \ (field ## _SHIFT)) & (field ## _MASK))
```
I'll drop it, thanks

GseoC commented:
Comments addressed

etienne-lms commented on core/include/drivers/stm32_rif.h:
```diff
@@ -64,24 +64,56 @@ struct rif_conf_data {
+ * Returns true if the given CID can't request the semaphore ownership,
+ * false otherwise
+ */
+static inline bool stm32_rif_semaphore_mode_incorrect(uint32_t cidcfgr,
```
Your're right, they looks slightly different but that are the same (but that they return an opposite value).

GseoC commented:
@etienne-lms , can I squash the review patches?

etienne-lms commented:
yes, please do.

GseoC commented:
Rebased and commits squashed

etienne-lms commented on core/drivers/firewall/stm32_rifsc.c:
```diff
@@ -0,0 +1,560 @@
+					    cidcfgr_offset);
+		risup->sec = (bool)(seccgfr & BIT(perih_offset));
+		risup->priv = (bool)(privcgfr & BIT(perih_offset));
+		risup->lock = (bool)(lockcfgr & BIT(perih_offset));
```
cast operations are not strictly needed

etienne-lms commented on core/drivers/firewall/stm32_rifsc.c:
```diff
@@ -0,0 +1,560 @@
+						  cidcfgr_offset,
+						  MAX_CID_SUPPORTED);
+		if (res) {
+			EMSG("Could not acquire semaphore for resource %u",
```
/%u"/%"PRIu32/

etienne-lms commented on core/drivers/firewall/stm32_rifsc.c:
```diff
@@ -0,0 +1,560 @@
+				((semcfgr & _SEMCR_SEMCID_MASK) >>
+				 _SEMCR_SEMCID_SHIFT) == RIF_CID1;
+
+		FMSG("RIF semaphore %s for ID: %u",
```
/%u"/%"PRIu32/

etienne-lms commented on core/drivers/firewall/stm32_rifsc.c:
```diff
@@ -0,0 +1,560 @@
+
+	if (((io_read32(rifsc_pdata.base + _RIFSC_RIMC_CR) &
+	     _RIFSC_RIMC_CR_TDCID_MASK)) == (RIF_CID1 << SCID_SHIFT))
+		*tdcid_state = true;
```
Prefer
```c
	if (((io_read32(rifsc_pdata.base + _RIFSC_RIMC_CR) &
	     _RIFSC_RIMC_CR_TDCID_MASK)) == (RIF_CID1 << SCID_SHIFT))
		*tdcid_state = true;
	else
		*tdcid_state = false;
```
or even
```c
	uint32_t tdcid = (io_read32(rifsc_pdata.base + _RIFSC_RIMC_CR) &
			  _RIFSC_RIMC_CR_TDCID_MASK) >> SCID_SHIFT;

	*tdcid_state = (tdcid == RIF_CID1);

etienne-lms commented on core/drivers/firewall/stm32_rifsc.c:
```diff
@@ -0,0 +1,560 @@
+		panic();
+	}
+
+	stm32_rifsc_glock_config(fdt, node, &rifsc_pdata);
```
test return code

GseoC commented on core/include/drivers/stm32_rif.h:
```diff
@@ -0,0 +1,224 @@
+ *					  semaphore ownership
+ *
+ * @cidcfgr: Value of the cidcfgr register
+ * @cid_to_check: CID to check (bitfield)
```
changing to cid_bf_to_check

GseoC commented on core/drivers/firewall/stm32_rifsc.c:
```diff
@@ -0,0 +1,560 @@
+		DMSG("No RISUP configuration in DT");
+		return TEE_ERROR_ITEM_NOT_FOUND;
+	}
+
```
Ok why not

GseoC commented:
Rebased with comments addressed. I've added a check in get_scid_mask() to forbid 0 as a valid number of CID supported

etienne-lms commented on core/drivers/firewall/stm32_rifsc.c:
```diff
@@ -0,0 +1,562 @@
+						  cidcfgr_offset,
+						  MAX_CID_SUPPORTED);
+		if (res) {
+			EMSG("Couldn't acquire semaphore for resource %u",
```
s/%u"/%"PRIu32/

etienne-lms commented on core/drivers/firewall/stm32_rifsc.c:
```diff
@@ -0,0 +1,562 @@
+						  cidcfgr_offset,
+						  MAX_CID_SUPPORTED);
+		if (res) {
+			EMSG("Couldn't release semaphore for resource %u",
```
s/%u"/%"PRIu32/

etienne-lms commented on core/include/dt-bindings/firewall/stm32mp25-rifsc.h:
```diff
@@ -0,0 +1,173 @@
+#define RIFSC_RIMC_MSEC_SHIFT		8
+#define RIFSC_RIMC_MPRIV_SHIFT		9
+#define RIFSC_RIMC_M_ID_SHIFT		16
+#define RIFSC_RIMC_ATTRx_SHIFT		0
```
This macro is not used, better remove its definition.

GseoC replied:
It is used lign 275 through _RIF_FLD_GET()

etienne-lms replied:
Right! my bad. I could not find it grepping the code but I forgot you defined these macros.

etienne-lms reviewed (commented):
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits
"drivers: firewall: add stm32_rif driver for common RIF features"
and "drivers: add stm32 RIFSC support"
with remaining review comments addressed.

GseoC commented:
Rebased with comments addressed and final tags applied. Thanks.
